### ISSUE TITLE
How to train the model with multi-class dataset

### ISSUE THREAD

**Issue by lpq29743** [alexander-rakhlin/CNN-for-Sentence-Classification-in-Keras#37]
Today, I tried to train the model using my own dataset. But disappointedly, it performs like having an error. To make my problem clearly, I use one example to train the model:
train_data:

> [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
> 5

test_data:

> [1, 13, 14, 15, 16, 17, 18, 19, 8, 20, 21, 22, 12]
> 4

vocabulary:

> 请问 1
> 京都 2
> 议定书 3
> 规定 4
> 几个 5
> 工业 6
> 国家 7
> 的 8
> 二氧化碳 9
> 排放量 10
> 限制 11
> ？ 12
> 金氏 13
> 世界纪录 14
> 中 15
> 最长 16
> （ 17
> 1650CM 18
> ） 19
> 发绣 20
> 作品 21
> 是 22

And the code that I had modified as follows.
Loading training data and testing data:

```
x_train, y_train, x_test, y_test = [], [], [], []
lines = open('data/train_data.txt', 'r').readlines()
for i in range(0, len(lines), 2):
    x_train.append(ast.literal_eval(lines[i]))
    y_train.append(int(lines[i + 1].replace('\n', '')))
lines = open('data/test_data.txt', 'r').readlines()
for i in range(0, len(lines), 2):
    x_test.append(ast.literal_eval(lines[i]))
    y_test.append(int(lines[i + 1].replace('\n', '')))
x_train = np.asarray(x_train)
y_train = np.asarray(y_train)
x_test = np.asarray(x_test)
y_test = np.asarray(y_test)
# (x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_words, start_char=None, oov_char=None, index_from=None)
```

Loading vocabulary:

```
vocabulary = {}
with open('data/data_info.txt', 'r') as f:
for line in f:
    content = line.strip().split()
    vocabulary[content[0]] = int(content[1])
# vocabulary = imdb.get_word_index()
vocabulary_inv = dict((v, k) for k, v in vocabulary.items())
vocabulary_inv[0] = "<PAD/>"
```

Loading word vectors:
```
# embedding_weights = train_word2vec(np.vstack((x_train, x_test)), vocabulary_inv, num_features=embedding_dim, min_word_count=min_word_count, context=context)
embedding_weights = {key: np.random.uniform(-0.25, 0.25, embedding_dim) for key, word in vocabulary_inv.items()}
```
                         
Run the project and it shows:

> Train on 1 samples, validate on 1 samples
> Epoch 1/10
>  - 0s - loss: 1.6577 - acc: 0.0000e+00 - val_loss: 0.9745 - val_acc: 0.0000e+00
> Epoch 2/10
>  - 0s - loss: 5.3148 - acc: 0.0000e+00 - val_loss: 0.8149 - val_acc: 0.0000e+00
> Epoch 3/10
>  - 0s - loss: -1.1496e+00 - acc: 0.0000e+00 - val_loss: 0.6260 - val_acc: 0.0000e+00
> Epoch 4/10
>  - 0s - loss: 1.4295 - acc: 0.0000e+00 - val_loss: 0.5458 - val_acc: 0.0000e+00
> Epoch 5/10
>  - 0s - loss: 1.8466 - acc: 0.0000e+00 - val_loss: 0.4188 - val_acc: 0.0000e+00
> Epoch 6/10
>  - 0s - loss: -1.1397e+00 - acc: 0.0000e+00 - val_loss: 0.2414 - val_acc: 0.0000e+00
> Epoch 7/10
>  - 0s - loss: -3.5465e-01 - acc: 0.0000e+00 - val_loss: -1.7151e-02 - val_acc: 0.0000e+00
> Epoch 8/10
>  - 0s - loss: -9.7948e-01 - acc: 0.0000e+00 - val_loss: -2.8196e-01 - val_acc: 0.0000e+00
> Epoch 9/10
>  - 0s - loss: 1.3985 - acc: 0.0000e+00 - val_loss: -5.5554e-01 - val_acc: 0.0000e+00
> Epoch 10/10
>  - 0s - loss: 3.3196 - acc: 0.0000e+00 - val_loss: -7.9620e-01 - val_acc: 0.0000e+00

**Comment by alexander-rakhlin** [comment]
From the train log it follows it is being trained on 1 sample, validated on 1 samples, which is not normal. Furthermore, for multi-class you would need to amend the model's output to softmax and adjust the targets.

**Comment by lpq29743** [comment]
Thank @alexander-rakhlin for answering my question. I using two samples just to clear my question, and the second suggestion is useful.
After checking my codes, I found some details should be noticed for multi-class classification.
1. Changing `y_train` and `y_test` for categorical_crossentropy
```
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)
```
2. Changing dense layer
```
model_output = Dense(num_of_classes, activation="softmax")(z)
```
3. Changing loss function
```
model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])
```
